lib/Dialect/FIRRTL/InstanceGraph.cpp

```diff
@@ -0,0 +1,93 @@
+//===- InstanceGraph.cpp - Instance Graph -----------------------*- C++ -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "circt/Dialect/FIRRTL/InstanceGraph.h"
+
+using namespace circt;
+using namespace firrtl;
+
+InstanceRecord *InstanceGraphNode::recordInstance(InstanceOp instance,
+                                                  InstanceGraphNode *target) {
+  auto *record = new InstanceRecord(instance, this, target);
+  moduleInstances.emplace_back(record);
+  return record;
+}
+
+void InstanceGraphNode::recordUse(InstanceRecord *record) {
+  moduleUses.push_back(record);
+}
+
+InstanceGraph::InstanceGraph(Operation *operation) {
+  auto circuitOp = cast<CircuitOp>(operation);
+
+  // We insert the top level module first in to the node map.  Getting the node
+  // here is enough to ensure that it is the first one added.
+  getOrAddNode(circuitOp.name());
+
+  for (auto &op : *circuitOp.getBody()) {
+    if (auto extModule = dyn_cast<FExtModuleOp>(op)) {
+      auto *currentNode = getOrAddNode(extModule.getName());
+      currentNode->module = extModule;
+    }
+    if (auto module = dyn_cast<FModuleOp>(op)) {
+      auto *currentNode = getOrAddNode(module.getName());
+      currentNode->module = module;
+      // Find all instance operations in the module body.
+      module.body().walk([&](InstanceOp instanceOp) {
```

Comment:
there is also getOps<Type> which will walk over one type.  Might be lighter weight than calling walk.

Reply:
I went with `walk()` so it recurses in to WhenOps, but it is pretty heavy weight.